# Homework 2

The goal of this Homework is to write Python code to answer each of the questions below. You should be able to answer these questions using Pandas only. You should not need to import any other packages. You should not use Excel or any other means to answer these questions. **Since questions have subquestions, it would be helpful if you use comments to clearly indicate which question you are answering**.

## Introduction
The Centers for Medicare & Medicaid Services ([CMS](cms.gov)) is a federal agency within the United States Department of Health and Human Services (HHS) that administers the **Medicare program** and works in partnership with state governments to administer Medicaid, the Children's Health Insurance Program (CHIP), and health insurance portability standards.
 - **Medicare** is a national health insurance program that provides health insurance for Americans aged 65 and over

The **Medicare Part D** program, also called the Medicare prescription drug benefit, is an optional United States federal-government program to help Medicare beneficiaries pay for self-administered prescription drugs through prescription drug insurance premiums.

CMS maintains a lot of [public info](https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Medicare-Provider-Charge-Data/Part-D-Prescriber) on prescription drugs that are prescribed by individual physicians and other health care **providers** and paid for under the **Medicare Part D** Prescription Drug Program.

I downloaded the **2017 Provider Summary table** and sampled it down from 1.1 million records (360 MB) to roughly 6000 records (2MB). If you would like to check out the original dataset, you can download it from this site https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Medicare-Provider-Charge-Data/PartD2017   

But for this assignment, you need to use the provided downsampled dataset `part_d_drug_prescriptions_sample2.csv`. If you would like, the data dictionary is available [here](https://www.cms.gov/Research-Statistics-Data-and-Systems/Statistics-Trends-and-Reports/Medicare-Provider-Charge-Data/Downloads/Prescriber_Methods.pdf). The applicable section is Table 2 in the Appendix.


In [1]:
import pandas as pd
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_rows', 200)


In [2]:
pd.set_option('display.max_columns', None)

## Question 1

Use Pandas to import **part_d_drug_prescriptions_sample2.csv** into a DataFrame, named `part_d`. How many rows and colums does the dataset have?

In [3]:
part_d = pd.read_csv("part_d_drug_prescriptions_sample2.csv")

In [4]:
part_d.shape

(5815, 84)

The dataset contains 5815 rows and 84 columns.


## Question 2
This question has two parts:
1. Display the first 6 rows of `part_d`.
2. Display the last 6 rows of `part_d`.

In [5]:
# Q2(1)
print("The first 6 rows of part_d:")
part_d.head(6)

The first 6 rows of part_d:


npi nppes_provider_last_org_name nppes_provider_first_name  \
0  1003000126                    ENKESHAFI                   ARDALAN   
1  1003012022                      CUDZILO                     COREY   
2  1003020041                       BAXTER                  MARGARET   
3  1003031212                       KAYALI                      FADI   
4  1003043175                       BENSON                   REBECCA   
5  1003054206                       SEILER                    WARREN   

  nppes_provider_mi nppes_credentials nppes_provider_gender nppes_entity_code  \
0               NaN              M.D.                     M                 I   
1               NaN              M.D.                     M                 I   
2               NaN              M.D.                     F                 I   
3               NaN              M.D.                     M                 I   
4                 L            D.D.S.                     F                 I   
5                 B                MD                     M                 I   

  nppes_provider_street1 nppes_provider_street2 nppes_provider_city  \
0           900 SETON DR                    NaN          CUMBERLAND   
1    2240 SUTHERLAND AVE              SUITE 103           KNOXVILLE   
2      1100 LONG POND RD              SUITE 250           ROCHESTER   
3     600 N CATTLEMEN RD              SUITE 200            SARASOTA   
4     6601 LYNDALE AVE S                    NaN           RICHFIELD   
5         2700 ROGERS DR              SUITE 101            HOMEWOOD   

   nppes_provider_zip5  nppes_provider_zip4 nppes_provider_state  \
0              21502.0               1854.0                   MD   
1              37919.0               2333.0                   TN   
2              14626.0               1177.0                   NY   
3              34232.0               6422.0                   FL   
4              55423.0               2477.0                   MN   
5              35209.0               2054.0                   AL   

  nppes_provider_country specialty_description description_flag  \
0                     US     Internal Medicine                S   
1                     US     Pulmonary Disease                S   
2                     US       Family Practice                S   
3                     US   Hematology-Oncology                S   
4                     US               Dentist                T   
5                     US       General Surgery                S   

  medicare_prvdr_enroll_status  total_claim_count  total_30_day_fill_count  \
0                            E                677                    695.5   
1                            E                370                    439.5   
2                            E               5661                  10090.3   
3                            E               1621                   2017.0   
4                            E                 28                     28.0   
5                            E                 37                     37.0   

   total_drug_cost  total_day_supply  bene_count ge65_suppress_flag  \
0         32639.57             14788       234.0                NaN   
1         80259.16             12220        82.0                NaN   
2        282062.15            288001       610.0                NaN   
3       2426817.07             51956       274.0                NaN   
4           193.32               204        20.0                  #   
5           529.13               242         NaN                NaN   

   total_claim_count_ge65  total_30_day_fill_count_ge65  total_drug_cost_ge65  \
0                   516.0                    532.533333              23174.78   
1                   288.0                    339.466667              63202.53   
2                  4684.0                   8713.533333             235746.07   
3                  1551.0                   1910.900000            2282413.42   
4                     NaN                 

In [6]:
# Q2(2)
print("The last 6 rows of part_d:")
part_d.tail(6)

The last 6 rows of part_d:


npi nppes_provider_last_org_name nppes_provider_first_name  \
5809  1992936330                       MANLEY               CHRISTOPHER   
5810  1992949150                         WANG                       AMY   
5811  1992961429                       PARKER                   MELINDA   
5812  1992969141                      BURKART                      NORA   
5813  1992980767                 TANAWUTTIWAT                  TANYANAN   
5814  1992994446                       PATETE                   MICHAEL   

     nppes_provider_mi nppes_credentials nppes_provider_gender  \
5809                 J                MD                     M   
5810                 M                MD                     F   
5811                 S              PA-C                     F   
5812                 E              M.D.                     F   
5813               NaN              M.D.                     F   
5814                 L          MD, FACS                     M   

     nppes_entity_code  nppes_provider_street1 nppes_provider_street2  \
5809                 I         333 COTTMAN AVE                    NaN   
5810                 I  3130 N COUNTY ROAD 25A                    NaN   
5811                 I          3209 S 23RD ST                STE 340   
5812                 I     1805 HENNEPIN AVE N                    NaN   
5813                 I         2500 N STATE ST                    NaN   
5814                 I          213 PALERMO PL                    NaN   

     nppes_provider_city  nppes_provider_zip5  nppes_provider_zip4  \
5809        PHILADELPHIA              19111.0               2434.0   
5810                TROY              45373.0               1337.0   
5811              TACOMA              98405.0               1602.0   
5812             GLENCOE              55336.0                  NaN   
5813             JACKSON              39216.0               4500.0   
5814              VENICE              34285.0               2821.0   

     nppes_provider_state nppes_provider_country  \
5809                   PA                     US   
5810                   OH                     US   
5811                   WA                     US   
5812                   MN                     US   
5813                   MS                     US   
5814                   FL                     US   

                   specialty_description description_flag  \
5809                   Pulmonary Disease                S   
5810                   Internal Medicine                S   
5811                 Physician Assistant                S   
5812                     General Surgery                S   
5813  Clinical Cardiac Electrophysiology                S   
5814                      Otolaryngology                S   

     medicare_prvdr_enroll_status  total_claim_count  total_30_day_fill_count  \
5809                            E                112                    115.1   
5810                            E                597                    619.1   
5811                            N                262                    264.8   
5812                            E                 56                     65.0   
5813                            E               1054                   1334.8   
5814                            E               2169                   3731.7   

      total_drug_cost  total_day_supply  bene_count ge65_suppress_flag  \
5809         17681.18              2792        44.0                NaN   
5810         27063.48             14427       204.0                NaN   
5811         12347.32              4718       136.0                  #   
5812           685.66              1285        26.0                  #   
5813         90812.31             39199       153.0                NaN   
5814         84032.09             96017       983.0                NaN   

      total_claim_count_ge65  total_30_day_fill_count_ge65  \
5809                    95.0                     98.133333   
5810             

## Question 3
This question has 4 parts:
1. Use the `info()` function to examine the DataFrame from a high level. Take a moment to look at the number of non-null values and the datatypes of each column. 
2. How many non-null values does the `npi` column have? (Use the output of the `info()` function to answer this question.)
3. What is the data type of the `npi` column? (Use the output of the `info()` function to answer this question.)
4. The purpose of this question is to understand the data you are looking at. The `npi` stands for National Provider Identifier. Please provide a one or two sentence description of the NPI - you may use a web search (google/wikipedia/etc...) to answer this question. 

In [7]:
# Q3(1)
part_d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5815 entries, 0 to 5814
Data columns (total 84 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   npi                               5815 non-null   int64  
 1   nppes_provider_last_org_name      5815 non-null   object 
 2   nppes_provider_first_name         5815 non-null   object 
 3   nppes_provider_mi                 4074 non-null   object 
 4   nppes_credentials                 5461 non-null   object 
 5   nppes_provider_gender             5815 non-null   object 
 6   nppes_entity_code                 5815 non-null   object 
 7   nppes_provider_street1            5815 non-null   object 
 8   nppes_provider_street2            2107 non-null   object 
 9   nppes_provider_city               5815 non-null   object 
 10  nppes_provider_zip5               5815 non-null   float64
 11  nppes_provider_zip4               5031 non-null   float64
 12  nppes_

- Q3(2). The npi column has 5815 non-null values.
- Q3(3). The data type of the npi column is int64(intiger).
- Q3(4). A National Provider Identifier(NPI) is a unique 10-digit identification number issued to health care providers in the United States by the Centers for Medicare and Medicaid Services.

## Question 4
Take a look at the `nppes_provider_state` column. Compute the number of unique values in this column using at least two different methods.

In [8]:
# Method 1
n = len(pd.unique(part_d['nppes_provider_state'])) 
print("Number of unique values in the nppes_provider_state column:", n)

Number of unique values in the nppes_provider_state column: 55


In [9]:
# Method 2
n = part_d.nppes_provider_state.nunique()  
print("Number of unique values in the nppes_provider_state column:", n)

Number of unique values in the nppes_provider_state column: 55


## Question 5
Let's take a look at the `specialty_description` column. Use the `value_counts()` method to determine the TOP 5 specialties in the dataset as well as their frequencies.

In [10]:
part_d['specialty_description'].value_counts()[:5]

Nurse Practitioner     785
Dentist                647
Internal Medicine      626
Family Practice        565
Physician Assistant    446
Name: specialty_description, dtype: int64

## Question 6
This question has four parts:
1. Define a variable `N` equal to the number of rows in the DataFrame, using the DataFrame `len` or `shape` methods.
2. Append the `cumsum()` function to the output of the `value_counts()` function on the `specialty_description` field  and divide the output by `N` to obtain the cumulative ratio to the total count. 
3. Create a boolean variable (eg a mask) which resolves to `True` if the cumulative sum is less than 75%, else `False`.
4. Use this boolean mask to filter the output of `value_counts()` to identify the records that contribute to the first 75% of all specialities.

In [11]:
# Q6(1)
N = len(part_d)
N

5815

In [12]:
# Q6(2)
specialty_cumulative_ratio = part_d['specialty_description'].value_counts().cumsum()/N
specialty_cumulative_ratio = specialty_cumulative_ratio.to_frame()
specialty_cumulative_ratio

specialty_description
Nurse Practitioner                                               0.134996
Dentist                                                          0.246260
Internal Medicine                                                0.353912
Family Practice                                                  0.451075
Physician Assistant                                              0.527773
Emergency Medicine                                               0.570765
Student in an Organized Health Care Education/T...               0.613758
Obstetrics & Gynecology                                          0.648667
Optometry                                                        0.677558
Psychiatry                                                       0.703697
General Surgery                                                  0.724506
Ophthalmology                                                    0.744282
Cardiology                                                       0.760791
Orthopedic Surgery                                               0.775752
Podiatry                                                         0.790714
Neurology                                                        0.803439
Gastroenterology                                                 0.814101
Dermatology                                                      0.824420
Psychiatry & Neurology                                           0.834394
Pediatric Medicine                                               0.843336
Urology                                                          0.852107
Hematology-Oncology                                              0.860533
Pulmonary Disease                                                0.868960
Otolaryngology                                                   0.876870
General Practice                                                 0.884609
Nephrology                                                       0.890628
Pharmacist                                                       0.896475
Endocrinology                                                    0.901806
Hospitalist                                                      0.906965
Anesthesiology                                                   0.912124
Physical Medicine and Rehabilitation                             0.917111
Rheumatology                                                     0.922098
Infectious Disease                                               0.926741
Radiation Oncology                                               0.930868
Allergy/ Immunology                                              0.934824
Oral Surgery (Dentist only)                                      0.938779
Plastic and Reconstructive Surgery                               0.942734
Neurosurgery                                                     0.946002
Interventional Cardiology                                        0.949269
Pain Management                                                  0.952365
Diagnostic Radiology                                             0.955288
Vascular Surgery                                                 0.958212
Medical Oncology                                                 0.961135
Certified Nurse Midwife                                          0.963887
Orthopaedic Surgery                                              0.966638
Specialist                                                       0.969218
Clinical Cardiac Electrophysiology                               0.971625
Certified Clinical Nurse Specialist                              0.974033
Neuropsychiatry                                                  0.975752
Interventional Pain Management                                   0.977472
Geriatric Medicine                                               0.979020
Critical Care (Intensivists)                                     0.980396
Plastic Surgery                                                  0.981771
Colorectal Surgery (Proctology)                         

In [13]:
# Q6(3) Create a boolean variable (eg a mask) which resolves to True if the cumulative sum is less than 75%, else False.
specialty_cumulative_ratio['mask'] = specialty_cumulative_ratio['specialty_description'] < 0.75
specialty_cumulative_ratio

specialty_description  \
Nurse Practitioner                                               0.134996   
Dentist                                                          0.246260   
Internal Medicine                                                0.353912   
Family Practice                                                  0.451075   
Physician Assistant                                              0.527773   
Emergency Medicine                                               0.570765   
Student in an Organized Health Care Education/T...               0.613758   
Obstetrics & Gynecology                                          0.648667   
Optometry                                                        0.677558   
Psychiatry                                                       0.703697   
General Surgery                                                  0.724506   
Ophthalmology                                                    0.744282   
Cardiology                                                       0.760791   
Orthopedic Surgery                                               0.775752   
Podiatry                                                         0.790714   
Neurology                                                        0.803439   
Gastroenterology                                                 0.814101   
Dermatology                                                      0.824420   
Psychiatry & Neurology                                           0.834394   
Pediatric Medicine                                               0.843336   
Urology                                                          0.852107   
Hematology-Oncology                                              0.860533   
Pulmonary Disease                                                0.868960   
Otolaryngology                                                   0.876870   
General Practice                                                 0.884609   
Nephrology                                                       0.890628   
Pharmacist                                                       0.896475   
Endocrinology                                                    0.901806   
Hospitalist                                                      0.906965   
Anesthesiology                                                   0.912124   
Physical Medicine and Rehabilitation                             0.917111   
Rheumatology                                                     0.922098   
Infectious Disease                                               0.926741   
Radiation Oncology                                               0.930868   
Allergy/ Immunology                                              0.934824   
Oral Surgery (Dentist only)                                      0.938779   
Plastic and Reconstructive Surgery                               0.942734   
Neurosurgery                                                     0.946002   
Interventional Cardiology                                        0.949269   
Pain Management                                                  0.952365   
Diagnostic Radiology                                             0.955288   
Vascular Surgery                                                 0.958212   
Medical Oncology                                                 0.961135   
Certified Nurse Midwife                                          0.963887   
Orthopaedic Surgery                                              0.966638   
Specialist                                                       0.969218   
Clinical Cardiac Electrophysiology                               0.971625   
Certified Clinical Nurse Specialist                              0.974033   
Neuropsychiatry                                                  0.975752   
Interventional Pain Management                                   0.977472   
Geriatric Medicine                                               0.979020   
Critical Care (Intensivists)                    

In [14]:
# Q6(4) Use this boolean mask to filter the output of value_counts() 
# to identify the records that contribute to the first 75% of all specialities.
specialty_cumulative_ratio[specialty_cumulative_ratio['mask'] == True]


specialty_description  \
Nurse Practitioner                                               0.134996   
Dentist                                                          0.246260   
Internal Medicine                                                0.353912   
Family Practice                                                  0.451075   
Physician Assistant                                              0.527773   
Emergency Medicine                                               0.570765   
Student in an Organized Health Care Education/T...               0.613758   
Obstetrics & Gynecology                                          0.648667   
Optometry                                                        0.677558   
Psychiatry                                                       0.703697   
General Surgery                                                  0.724506   
Ophthalmology                                                    0.744282   

                                                    mask  
Nurse Practitioner                                  True  
Dentist                                             True  
Internal Medicine                                   True  
Family Practice                                     True  
Physician Assistant                                 True  
Emergency Medicine                                  True  
Student in an Organized Health Care Education/T...  True  
Obstetrics & Gynecology                             True  
Optometry                                           True  
Psychiatry                                          True  
General Surgery                                     True  
Ophthalmology                                       True

## Question 7:
Take a look at these two columns: `total_claim_count` and `opioid_claim_count`. Define a new column that stores the ratio of the two, eg:
```python
part_d['opioid_claim_rate'] = part_d['opioid_claim_count']/part_d['total_claim_count']
```
Let us define **high opioid providers** as those who have an `opioid_claim_rate` that is greater than or equal to 50%. This question has three parts:
1. Run the describe method on the `opioid_claim_rate` column and note the statistics.
2. Use TWO different methods to calculate how many **high opioid providers** there are in the dataset. (Hint: you can use the sum of a boolean variable for one method and the size of a filtered dataset for the other method).
3. Use TWO different methods to identify the top 3 states that have the highest number of **high opioid providers**. (Hint: you can use value_counts for one method and groupby for the other method). 

In [15]:
# Q7(1)
part_d['opioid_claim_rate'] = part_d['opioid_claim_count']/part_d['total_claim_count']
part_d.head()

npi nppes_provider_last_org_name nppes_provider_first_name  \
0  1003000126                    ENKESHAFI                   ARDALAN   
1  1003012022                      CUDZILO                     COREY   
2  1003020041                       BAXTER                  MARGARET   
3  1003031212                       KAYALI                      FADI   
4  1003043175                       BENSON                   REBECCA   

  nppes_provider_mi nppes_credentials nppes_provider_gender nppes_entity_code  \
0               NaN              M.D.                     M                 I   
1               NaN              M.D.                     M                 I   
2               NaN              M.D.                     F                 I   
3               NaN              M.D.                     M                 I   
4                 L            D.D.S.                     F                 I   

  nppes_provider_street1 nppes_provider_street2 nppes_provider_city  \
0           900 SETON DR                    NaN          CUMBERLAND   
1    2240 SUTHERLAND AVE              SUITE 103           KNOXVILLE   
2      1100 LONG POND RD              SUITE 250           ROCHESTER   
3     600 N CATTLEMEN RD              SUITE 200            SARASOTA   
4     6601 LYNDALE AVE S                    NaN           RICHFIELD   

   nppes_provider_zip5  nppes_provider_zip4 nppes_provider_state  \
0              21502.0               1854.0                   MD   
1              37919.0               2333.0                   TN   
2              14626.0               1177.0                   NY   
3              34232.0               6422.0                   FL   
4              55423.0               2477.0                   MN   

  nppes_provider_country specialty_description description_flag  \
0                     US     Internal Medicine                S   
1                     US     Pulmonary Disease                S   
2                     US       Family Practice                S   
3                     US   Hematology-Oncology                S   
4                     US               Dentist                T   

  medicare_prvdr_enroll_status  total_claim_count  total_30_day_fill_count  \
0                            E                677                    695.5   
1                            E                370                    439.5   
2                            E               5661                  10090.3   
3                            E               1621                   2017.0   
4                            E                 28                     28.0   

   total_drug_cost  total_day_supply  bene_count ge65_suppress_flag  \
0         32639.57             14788       234.0                NaN   
1         80259.16             12220        82.0                NaN   
2        282062.15            288001       610.0                NaN   
3       2426817.07             51956       274.0                NaN   
4           193.32               204        20.0                  #   

   total_claim_count_ge65  total_30_day_fill_count_ge65  total_drug_cost_ge65  \
0                   516.0                    532.533333              23174.78   
1                   288.0                    339.466667              63202.53   
2                  4684.0                   8713.533333             235746.07   
3                  1551.0                   1910.900000            2282413.42   
4                     NaN                           NaN                   NaN   

   total_day_supply_ge65 bene_count_ge65_suppress_flag  bene_count_ge65  \
0                11428.0                           NaN            182.0   
1                 9472.0                           NaN             58.0   
2               249616.0                           NaN            532.0   
3                49017.0                           NaN            258.0   
4                    NaN                             #              NaN   

  brand_suppress_

In [16]:
# note the statistics
part_d['opioid_claim_rate'].describe()

count    4225.000000
mean        0.097227
std         0.164038
min         0.000000
25%         0.000000
50%         0.022523
75%         0.113238
max         1.000000
Name: opioid_claim_rate, dtype: float64

In [17]:
# Q7(2) High opioid providers as those who have an opioid_claim_rate that is greater than or equal to 50%.
# Use TWO different methods to calculate how many high opioid providers there are in the dataset.
high_opioid_provider = part_d['opioid_claim_rate'] >= 0.5
high_opioid_provider

0       False
1       False
2       False
3       False
4       False
        ...  
5810    False
5811    False
5812    False
5813    False
5814    False
Name: opioid_claim_rate, Length: 5815, dtype: bool

In [18]:
# Q7(2) Method 1
sum(high_opioid_provider)

209

In [19]:
# Q7(2) Method 2
part_d[high_opioid_provider].shape

(209, 85)

There are 209 high opioid providers in the dataset.

In [20]:
# Q7(3) Use TWO different methods to identify the top 3 states that have the highest number of high opioid providers. 
# Method 1 value_counts
high_opioid = part_d[part_d['opioid_claim_rate'] >= 0.5]
high_opioid['nppes_provider_state'].value_counts()[:3]

CA    19
FL    15
MI    12
Name: nppes_provider_state, dtype: int64

In [21]:
# Q7(3) Method 2 groupby
high_opioid_state = high_opioid.groupby('nppes_provider_state').count().sort_values('opioid_claim_rate',ascending=False)[:3]
high_opioid_state['opioid_claim_rate']

nppes_provider_state
CA    19
FL    15
NY    12
Name: opioid_claim_rate, dtype: int64

## Question 8:
Using the `opioid_claim_rate` column computed in the previous question:

1. Create a new DataFrame `g` that aggregates the `nppes_provider_gender` column using the `groupby` method and the `size` function to output the gender breakdown in the dataset. You can use this code to answer this question:
```python
g = part_d.groupby('nppes_provider_gender').size().reset_index()
```
2. Use the `rename` method to rename the columns of DataFrame `g` to *gender* and *count*.
3. Now compute the aggregation of the dataset by gender but filtered to the **high opioid providers**. Name the DataFrame `g2`. Rename the columns to *gender* and *high_opioid_count*.
4. Join the DataFrames g and g2 using the gender column as the key. The final result should look like this:
```
gender	count	high_opioid_count
	F	2458	62
	M	3357	147
```
5. Compute the ratio of `high_opioid_count` to `count` for each row and note the difference

In [22]:
# Q8(1)
g = part_d.groupby('nppes_provider_gender').size().reset_index()
g

nppes_provider_gender     0
0                     F  2458
1                     M  3357

In [23]:
# Q8(2)
g.rename(columns = {'nppes_provider_gender':'gender', 0:'count'}, inplace=True)
g

gender  count
0      F   2458
1      M   3357

In [24]:
# Q8(3) Now compute the aggregation of the dataset by gender but filtered to the high opioid providers. 
# Name the DataFrame g2. Rename the columns to gender and high_opioid_count.
g2 = part_d[part_d['opioid_claim_rate']>=0.5].groupby('nppes_provider_gender').size().reset_index()
g2.rename(columns = {'nppes_provider_gender':'gender', 0:'high_opioid_count'},inplace = True)
g2

gender  high_opioid_count
0      F                 62
1      M                147

In [25]:
# Q8(4) Join the DataFrames g and g2 using the gender column as the key.
new_g = pd.merge(g,g2,on = 'gender')
new_g

gender  count  high_opioid_count
0      F   2458                 62
1      M   3357                147

In [26]:
# Q8(5) Compute the ratio of high_opioid_count to count for each row and note the difference
g2['high_opioid_ratio'] = g2['high_opioid_count']/sum(g2['high_opioid_count'])
g2

gender  high_opioid_count  high_opioid_ratio
0      F                 62           0.296651
1      M                147           0.703349

Female opioid provider has a high opioid ratio at 29.67% while male opioid provider has a high opioid ratio at 70.33%.

## Question 9:
The Office of Inspector General ([OIG](https://oig.hhs.gov/)) has the authority to **exclude** individuals and entities from Federally funded health care programs. To be excluded is quite serious, as you can see from their website:

>The OIG maintains a list of all currently excluded individuals and entities called the List of Excluded Individuals/Entities (LEIE). Anyone who hires an individual or entity on the LEIE may be subject to civil monetary penalties (CMP). To avoid CMP liability, health care entities should routinely check the list to ensure that new hires and current employees are not on it.

This list is publicaly available, [here](https://oig.hhs.gov/exclusions/exclusions_list.asp). I have downloaded this file in its entirety (in December 2020) and have stripped it down to the columns we need.

Load the provided file into a DataFrame, named `exclusions`:

```python
exclusions = pd.read_csv('physicians_exclusions_sample3.csv')
```
This question has a number of parts:

1. Print the number of rows and columns in the exclusions dataset.
2. Display the first 6 rows the DataFrame. 
3. Display 3 random rows of the DataFrame using the `sample` method.
4. Use the `nunique` method to determine the number of unique NPIs. Compare this to the size of the dataset - what do you observe?
5. Are there any NULL `NPI` values?
6. What is the most frequent `NPI` value?
7. Apply the `unique` method to the NPI column and convert the outcome to a list of NPIs (hint use the `tolist` method). Store the output in a variable called `npis`. What is the length of `npis`?
8. Apply the `isin` method to `part_d` to tag the rows whose NPI is in the list of excluded NPIs. You can use the following code. How many records are there in `excluded1`?
```python
excluded1 = part_d[part_d['npi'].isin(npis)]
```

In [27]:
# Q9(1)
exclusions = pd.read_csv('physicians_exclusions_sample3.csv')
exclusions.shape

(69858, 4)

The exclusions dataset has 69858 rows and 4 columns.

In [28]:
# Q9(2) Display the first 6 rows the DataFrame.
print("The first 6 rows of the DataFrame: ")
exclusions.head(6)

The first 6 rows of the DataFrame: 


LASTNAME    FIRSTNAME STATE  NPI
0   AALDERS        NANCY    IL    0
1  AALTONEN     NICKOLAS    WA    0
2     AAMIR     MUHAMMAD    MI    0
3     AARON        ALINA    AZ    0
4     AARON  CHRISTOPHER    PA    0
5     AARON      DEBORAH    TX    0

In [29]:
# Q9(3) Display 3 random rows of the DataFrame using the sample method.
exclusions.sample(n=3)

LASTNAME FIRSTNAME STATE  NPI
52898      RIVERA  KATHRINE    NV    0
20283  FONTANELLA    DANIEL    NJ    0
21239       FULTZ      JEAN    NY    0

In [30]:
# Q9(4) Use the nunique method to determine the number of unique NPIs. 
# Compare this to the size of the dataset - what do you observe?
exclusions['NPI'].nunique()

5200

The dataset has 69,858 rows while NPI only has 5,200 unique values. The NPI variable has a high repetitive rate.

In [31]:
# Q9(5) Are there any NULL NPI values?
exclusions['NPI'].isnull().any().any()

False

There is no null NPI values.

In [32]:
# Q9(6) What is the most frequent NPI value?
exclusions['NPI'].value_counts()[:1]

0    64621
Name: NPI, dtype: int64

The most frequent NPI value is 0.

In [33]:
# Q9(7) Apply the unique method to the NPI column and convert the outcome to a list of NPIs 
# Store the output in a variable called npis. What is the length of npis?
npis = exclusions['NPI'].unique().tolist()
len(npis)

5200

In [34]:
# Q9(8) Apply the isin method to part_d to tag the rows whose NPI is in the list of excluded NPIs. 
# You can use the following code. How many records are there in excluded1?
excluded1 = part_d[part_d['npi'].isin(npis)]
excluded1

npi nppes_provider_last_org_name nppes_provider_first_name  \
3814  1659506111                      FRANCIS                   DESMOND   
4481  1770505612                         YANG                  HARRISON   

     nppes_provider_mi nppes_credentials nppes_provider_gender  \
3814               NaN                MD                     M   
4481                 Y  M.D., FACP, FACC                     M   

     nppes_entity_code   nppes_provider_street1 nppes_provider_street2  \
3814                 I  1801 BLACK RIVER BLVD N                    NaN   
4481                 I    845 MCARTHUR ST STE D                    NaN   

     nppes_provider_city  nppes_provider_zip5  nppes_provider_zip4  \
3814                ROME              13440.0               2427.0   
4481          MANCHESTER              37355.0               2365.0   

     nppes_provider_state nppes_provider_country specialty_description  \
3814                   NY                     US     Internal Medicine   
4481                   TN                     US     Internal Medicine   

     description_flag medicare_prvdr_enroll_status  total_claim_count  \
3814                S                            E               6383   
4481                S                            E              11429   

      total_30_day_fill_count  total_drug_cost  total_day_supply  bene_count  \
3814                  11548.7        540038.11            335808       352.0   
4481                  19489.1        785476.46            523948       463.0   

     ge65_suppress_flag  total_claim_count_ge65  total_30_day_fill_count_ge65  \
3814                NaN                  3740.0                   7339.933333   
4481                NaN                  6963.0                  11919.900000   

      total_drug_cost_ge65  total_day_supply_ge65  \
3814             294157.56               216316.0   
4481             402694.74               315030.0   

     bene_count_ge65_suppress_flag  bene_count_ge65 brand_suppress_flag  \
3814                           NaN            223.0                 NaN   
4481                           NaN            258.0                 NaN   

      brand_claim_count  brand_drug_cost generic_suppress_flag  \
3814             1054.0        397146.06                   NaN   
4481             1949.0        544147.47                   NaN   

      generic_claim_count  generic_drug_cost other_suppress_flag  \
3814               5254.0          140210.01                 NaN   
4481               9407.0          239442.90                 NaN   

      other_claim_count  other_drug_cost mapd_suppress_flag  mapd_claim_count  \
3814               75.0          2682.04                NaN            3135.0   
4481               73.0          1886.09                NaN            4201.0   

      mapd_drug_cost pdp_suppress_flag  pdp_claim_count  pdp_drug_cost  \
3814       266997.41               NaN           3248.0      273040.70   
4481       360517.57               NaN           7228.0      424958.89   

     lis_suppress_flag  lis_claim_count  lis_drug_cost nonlis_suppress_flag  \
3814               NaN           3532.0       340452.5                  NaN   
4481               NaN           8802.0       663165.5                  NaN   

      nonlis_claim_count  nonlis_drug_cost  opioid_claim_count  \
3814              2851.0         199585.61               574.0   
4481              2627.0         122310.96              1365.0   

      opioid_drug_cost  opioid_day_supply  opioid_bene_count  \
3814          81990.52            15641.0               64.0   
4481          25727.80            32105.0              228.0   

      opioid_prescriber_rate  la_opioid_claim_count  la_opioid_drug_cost  \
3814                    8.99                  254.0             78100.59   
4481                   11.94                    NaN                  NaN   

      la_opioid_day_supply  la_opioid_bene_count  la_opioid_prescriber_rate  \
3814                8745.0     

There are 2 records in excluded1.

## Question 10
In question 9 we searched for records in the part_d DataFrame whose NPI was in the exclusions dataset. Matching across the NPI yielded only 2 excluded providers. This seems suspicious. The problem is probably due to the fact that so many of the NPIs in the excluded dataset are marked as `0`. Perhaps we can do better by matching across first and last names. Let's do that in these steps.
1. As a sanity check, let's first do a merge between the two datasets on NPI. Your results should confirm the result obtained in question 9. For convenience, define this list of columns:
```python
cols = [
    'npi', 
    'nppes_provider_first_name', 
    'nppes_provider_last_org_name', 
    'nppes_provider_state', 
    'NPI',
    'FIRSTNAME',
    'LASTNAME',
    'STATE',
    'opioid_claim_rate'
]
```
Perform the merge and display the output for the columns in `cols`. You can use the following code to answer this question:
```python
excluded2 = pd.merge(left = part_d, right = exclusions, how = 'inner', 
         left_on = ['npi'],
         right_on = ['NPI']
        )
excluded2[cols]
```
Your results should match the results from Question 9.


2. Since we are going to match across names, let's first convert the applicable columns to upper case and strip out any leading or trailing spaces. You can use the following code to answer this question but you'll need to apply it to the remaining first and last name columns, eg 'nppes_provider_first_name', 'nppes_provider_last_org_name', 'FIRSTNAME' and 'LASTNAME'
```python
part_d['nppes_provider_first_name'] = part_d['nppes_provider_first_name'].str.upper().str.strip()
```

3. Run another inner join across both datasets using first and last names. My result had 435 rows. Please take a look at the dataset filtering on the columns in `cols` as done above. What do you notice? 


4. In order to conduct a better quality match, include the state in your join keys. You should now be matching on first name, last name and state. My result had 22 records. I feel more confident in this match since the names and state lines up. You'll notice that some of the providers have 0 `opioid_claim_rate`. This is not that surprising as there are many different kinds of behaviors that a provider can engage in in order to get on the exclusisons list. 


In [35]:
cols = [
    'npi', 
    'nppes_provider_first_name', 
    'nppes_provider_last_org_name', 
    'nppes_provider_state', 
    'NPI',
    'FIRSTNAME',
    'LASTNAME',
    'STATE',
    'opioid_claim_rate'
]

In [36]:
excluded2 = pd.merge(left = part_d, right = exclusions, how = 'inner', 
      left_on = ['npi'],
      right_on = ['NPI']
     )
excluded2[cols]

npi nppes_provider_first_name nppes_provider_last_org_name  \
0  1659506111                   DESMOND                      FRANCIS   
1  1770505612                  HARRISON                         YANG   

  nppes_provider_state         NPI FIRSTNAME LASTNAME STATE  opioid_claim_rate  
0                   NY  1659506111   DESMOND  FRANCIS    NY           0.089926  
1                   TN  1770505612  HARRISON     YANG    TN           0.119433

In [37]:
# Q10(2) 
part_d['nppes_provider_first_name'] = part_d['nppes_provider_first_name'].str.upper().str.strip()
part_d['nppes_provider_last_org_name'] = part_d['nppes_provider_last_org_name'].str.upper().str.strip()
exclusions['FIRSTNAME'] = exclusions['FIRSTNAME'].str.upper().str.strip()
exclusions['LASTNAME'] = exclusions['LASTNAME'].str.upper().str.strip()

In [38]:
# Q10(3)
excluded3 = pd.merge(left = part_d,right = exclusions,how='inner',
                 left_on = ['nppes_provider_first_name','nppes_provider_last_org_name'], 
                 right_on = ['FIRSTNAME','LASTNAME'])
excluded3[cols]

npi nppes_provider_first_name nppes_provider_last_org_name  \
0    1003153628                   RICHARD                      COLEMAN   
1    1003290974                   MARILYN                        JONES   
2    1003811837                      MARK                     ROBINSON   
3    1346286507                      MARK                     ROBINSON   
4    1003868787                   DOUGLAS                       NELSON   
..          ...                       ...                          ...   
430  1982648333                   TIMOTHY                       MILLER   
431  1982668497                    ANDREW                     JIMERSON   
432  1992142632                   JESSICA                     MARSHALL   
433  1992786305                      JOHN                          FOX   
434  1992905616                    SHARON                          LEE   

    nppes_provider_state         NPI FIRSTNAME  LASTNAME STATE  \
0                     TX           0   RICHARD   COLEMAN    CA   
1                     IN           0   MARILYN     JONES    AR   
2                     OH  1427112325      MARK  ROBINSON    NH   
3                     CA  1427112325      MARK  ROBINSON    NH   
4                     IL           0   DOUGLAS    NELSON    IA   
..                   ...         ...       ...       ...   ...   
430                   WV           0   TIMOTHY    MILLER    UT   
431                   GA  1427056076    ANDREW  JIMERSON    OH   
432                   OH           0   JESSICA  MARSHALL    FL   
433                   KS           0      JOHN       FOX    AL   
434                   CA           0    SHARON       LEE    TX   

     opioid_claim_rate  
0             0.047162  
1             0.017435  
2             0.000000  
3             0.000000  
4             0.196429  
..                 ...  
430                NaN  
431           0.175676  
432           0.083333  
433           0.048907  
434                NaN  

[435 rows x 9 columns]

I noticed that even though first and last name are the same, addresses are different, indicating that these are not the same person.

In [39]:
# Q10(4)
excluded4 = pd.merge(left = part_d,right = exclusions,how='inner',
                 left_on = ['nppes_provider_first_name','nppes_provider_last_org_name','nppes_provider_state'], 
                 right_on = ['FIRSTNAME','LASTNAME','STATE'])
excluded4[cols]

npi nppes_provider_first_name nppes_provider_last_org_name  \
0   1013070861                   CHARLES                       HARRIS   
1   1053349506                      JOSE                       TORRES   
2   1144278052                    ROBERT                       TAYLOR   
3   1144382672                      MARY                        SMITH   
4   1184608903                    ROBERT                          LEE   
5   1205807864                    ROBERT                         ROTH   
6   1205839776                    MARTHA                     MARTINEZ   
7   1275501629                    ROBERT                        JONES   
8   1346213907                   STEPHEN                          LEE   
9   1346223179                   RICHARD                        SMITH   
10  1356315410                       DUC                       NGUYEN   
11  1356408017                    ROBERT                     GONZALES   
12  1427333186                   REBECCA                       MARTIN   
13  1518012954                     DAVID                     WILLIAMS   
14  1528326428                   WILLIAM                         LANE   
15  1538442017                  JENNIFER                        ORTIZ   
16  1659506111                   DESMOND                      FRANCIS   
17  1700185345                      ALAN                         KATZ   
18  1770505612                  HARRISON                         YANG   
19  1871955906                    CARMEN                        LOPEZ   
20  1942309596                 MARGARITA                    RODRIGUEZ   
21  1942592324                  JENNIFER                        SMITH   

   nppes_provider_state         NPI  FIRSTNAME   LASTNAME STATE  \
0                    TX           0    CHARLES     HARRIS    TX   
1                    FL           0       JOSE     TORRES    FL   
2                    CA           0     ROBERT     TAYLOR    CA   
3                    MS           0       MARY      SMITH    MS   
4                    IL           0     ROBERT        LEE    IL   
5                    CA  1700869864     ROBERT       ROTH    CA   
6                    TX           0     MARTHA   MARTINEZ    TX   
7                    CA           0     ROBERT      JONES    CA   
8                    PA           0    STEPHEN        LEE    PA   
9                    IL           0    RICHARD      SMITH    IL   
10                   CA           0        DUC     NGUYEN    CA   
11                   CA           0     ROBERT   GONZALES    CA   
12                   TX           0    REBECCA     MARTIN    TX   
13                   CA           0      DAVID   WILLIAMS    CA   
14                   MA           0    WILLIAM       LANE    MA   
15                   NY           0   JENNIFER      ORTIZ    NY   
16                   NY  1659506111    DESMOND    FRANCIS    NY   
17                   NY           0       ALAN       KATZ    NY   
18                   TN  1770505612   HARRISON       YANG    TN   
19                   FL           0     CARMEN      LOPEZ    FL   
20                   FL           0  MARGARITA  RODRIGUEZ    FL   
21                   NC           0   JENNIFER      SMITH    NC   

    opioid_claim_rate  
0            0.000000  
1            0.040196  
2            0.000000  
3            0.000000  
4            0.172285  
5                 NaN  
6                 NaN  
7            0.034858  
8            0.000000  
9            0.000000  
10           0.727273  
11           0.000000  
12           0.000000  
13           0.281746  
14           0.261905  
15           0.000000  
16           0.089926  
17           0.000000  
18           0.119433  
19           0.000000  
20                NaN  
21           0.036195